recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +317,384 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
         return circuit.transform_qubits(lambda q: self.mapping[q])
 
 
+class DynamicLookAheadHeuristicCircuitTransformer(CircuitTransformer):
+    """Optimizer that transforms a circuit to satify a device's constraints.
+
+    This implements the initial mapping algorithm and the SWAP update algorithm
+    proposed by the paper "A Dynamic Look-Ahead Heuristic for the Qubit Mapping
+    Problem of NISQ Computer:
+    https://ieeexplore.ieee.org/abstract/document/8976109.
+
+    Reference:
+    P. Zhu, Z. Guan and X. Cheng, "A Dynamic Look-Ahead Heuristic for the
+    Qubit Mapping Problem of NISQ Computers," in IEEE Transactions on Computer-
+    Aided Design of Integrated Circuits and Systems, vol. 39, no. 12, pp. 4721-
+    4735, Dec. 2020, doi: 10.1109/TCAD.2020.2970594.
+    """
+    def __init__(self, device: cirq.Device):
+        super().__init__()
+        self.device = device
+
+    def build_physical_qubits_graph(
+        self,
+    ) -> Dict[cirq.GridQubit, List[cirq.GridQubit]]:
+        """Returns an adjacency graph of physical qubits of the device.
+
+        Each edge is bidirectional, and represents a valid two-qubit gate.
+        """
+        g = defaultdict(list)
+        for q in self.device.qubit_set():
+            neighbors = [n for n in q.neighbors() if n in self.device.qubit_set()]
+            for n in neighbors:
+                g[q].append(n)
+        return g
+
+    def get_least_connected_qubit(
+        self,
+        g: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+        component: Deque[cirq.Qid],
+    ) -> cirq.Qid:
+        """Returns the least connected qubit.
+
+        Args:
+          g: A logical qubits graph.
+          component: A deque of qubits belonging to the same component.
+        """
+        return min(component, key=lambda q: len(g[q]))
+
+    def build_logical_qubits_graph(
+        self,
+        circuit: cirq.Circuit,
+    ) -> Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]]:
+        """Returns an adjacency graph of logical qubits of the circuit.
+
+        Uses the heuristic of adding an edge between the nodes of each
+        disjoint component that are least connected if the graph contains more
+        than one connected component.
+
+        Each edge is a tuple containing an adjacent node and the index of the
+        moment at which the operation occurs.
+
+        Arg:
+          circuit: The circuit from which to build a logical qubits graph.
+        """
+        g = defaultdict(list)
+        moment_index = 0
+
+        # Build an adjacency graph based on the circuit.
+        for i, m in enumerate(circuit):
+            moment_index = i
+            for op in m:
+                if len(op.qubits) == 1:
+                    q = op.qubits[0]
+                    if q not in g:
+                        g[q] = []
+                if len(op.qubits) == 2:
```

Comment:
Should we change this to `elif len(op.qubits) == 2` and then add an else that handles len(op.qubits) > 2 (possibly by throwing a ValueError?  It might be a good idea to make sure we skip measurement gates properly too.

Reply:
sg, updated to throw a `ValueError` when `len(ops.qubits) > 2`; also added a check to skip measurement gates.